osdconfig/api.go

```diff
@@ -0,0 +1,275 @@
+	if manager.logLevel == LOG_INFO_AND_ERRORS {
+		logrus.Info(args...)
+	}
+}
+
+// GetClusterConf retrieves cluster level data from kvdb
+func (manager *configManager) GetClusterConf() (*ClusterConfig, error) {
+	// get json from kvdb and unmarshal into config
+	kvPair, err := manager.cc.Get(filepath.Join(base_key, cluster_key))
+	if err != nil {
+		return nil, err
+	}
+
+	if kvPair == nil {
+		return nil, errors.New("no data fetched from kvdb store")
+	}
+
+	config := new(ClusterConfig)
+	if err := json.Unmarshal(kvPair.Value, config); err != nil {
+		return nil, err
+	}
+
+	return config, nil
+}
+
+// SetClusterConf sets cluster config in kvdb
+func (manager *configManager) SetClusterConf(config *ClusterConfig) error {
+	if config == nil {
+		return errors.New("input is nil")
+	}
+
+	// push into kvdb
+	_, err := manager.cc.Put(filepath.Join(base_key, cluster_key), config, 0)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetNodeConf retrieves node config data using nodeID
+func (manager *configManager) GetNodeConf(nodeID string) (*NodeConfig, error) {
+	if len(nodeID) == 0 {
+		return nil, errors.New("node id input is empty")
+	}
+
+	// get json from kvdb and unmarshal into config
+	kvPair, err := manager.cc.Get(getNodeKeyFromNodeID(nodeID))
+	if err != nil {
+		return nil, err
+	}
+
+	if kvPair == nil {
+		return nil, errors.New("no data fetched from kvdb store")
+	}
+
+	config := new(NodeConfig)
+	if err := json.Unmarshal(kvPair.Value, config); err != nil {
+		return nil, err
+	}
+
+	return config, nil
+}
+
+// SetNodeConf sets node config data in kvdb
+func (manager *configManager) SetNodeConf(config *NodeConfig) error {
+	if config == nil {
+		return errors.New("input is nil")
+	}
+
+	if len(config.NodeId) == 0 {
+		return errors.New("node id in input is empty")
+	}
+
+	// push into kvdb
+	_, err := manager.cc.Put(getNodeKeyFromNodeID(config.NodeId), config, 0)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// RegisterCB registers callback functions
+// callback to be registered is expected to be a service delivering a channel to write on
+func (manager *configManager) RegisterCB(name string, opt interface{},
+	cb func(ctx context.Context,
+		opt interface{}) (chan<- *DataToCallback, <-chan *DataFromCallback)) error {
+
+	// obtain lock since registering a callback changes the state of manager object
+	manager.Lock()
+	defer manager.Unlock()
+
+	// callbacks are registered by the "name" string. only one per name
+	if _, present := manager.cb[name]; !present {
+		cbi := new(callbackData)
+		cbi.f = cb
+		cbi.opt = opt
+		manager.cb[name] = cbi
+	} else {
+		return errors.New("callback already registered with this name: " + name)
+	}
+
+	return nil
+}
+
+// Run loops over all registered callbacks and executes them.
+// It returns a channel on which you only read and it contains the errors.
+func (manager *configManager) Run(wd *DataToCallback) <-chan error {
+	manager.Lock()
```

Comment:
its probably better to have lock for members instead of the entire struct

Reply:
i would like to protect both the data state and execution state associated with an instance. A single lock is easier to manage.